src/hotspot/share/opto/superword.cpp

```diff
@@ -3615,3 +3030,279 @@ bool SuperWord::same_origin_idx(Node* a, Node* b) const {
+      shift_count->set_req(1, same_input_vtn);
+      return shift_count;
+    } else {
+      // Replicate the scalar same_input to every vector element.
+      const Type* element_type = _vloop_analyzer.types().velt_type(p0);
+      if (index == 2 && VectorNode::is_scalar_rotate(p0) && element_type->isa_long()) {
+        // Scalar rotate has int rotation value, but the scalar rotate expects longs.
+        assert(same_input->bottom_type()->isa_int(), "scalar rotate expects int rotation");
+        VTransformNode* conv = new (_graph.arena()) VTransformConvI2LNode(_graph);
+        conv->set_req(1, same_input_vtn);
+        same_input_vtn = conv;
+      }
+      VTransformNode* replicate = new (_graph.arena()) VTransformReplicateNode(_graph, pack->size(), element_type);
+      replicate->set_req(1, same_input_vtn);
+      return replicate;
+    }
+  }
+
+  // The input is neither a pack not a same_input node. SuperWord::profitable does not allow
+  // any other case. In the future, we could insert a PackNode.
+#ifdef ASSERT
+  tty->print_cr("\nSuperWordVTransformBuilder::get_vtnode_vector_input_at_index: index=%d", index);
+  pack->dump();
+  assert(false, "Pack input was neither a pack nor a same_input node");
+#endif
+  ShouldNotReachHere();
+}
+
+VTransformNode* SuperWordVTransformBuilder::get_vtnode_or_wrap_as_input_scalar(Node* n) {
+  VTransformNode* vtn = get_vtnode_or_null(n);
+  if (vtn != nullptr) { return vtn; }
+
+  assert(!_vloop.in_bb(n), "only nodes outside the loop can be input nodes to the loop");
+  vtn = new (_graph.arena()) VTransformInputScalarNode(_graph, n);
+  set_vtnode(n, vtn);
+  return vtn;
+}
+
+void SuperWordVTransformBuilder::set_req_with_vector(const Node_List* pack, VTransformNode* vtn, VectorSet& vtn_dependencies, int j) {
+  VTransformNode* req = get_vtnode_vector_input_at_index(pack, j);
+  vtn->set_req(j, req);
+  vtn_dependencies.set(req->_idx);
+}
+
+void SuperWordVTransformBuilder::set_all_req_with_scalars(Node* n, VTransformNode* vtn, VectorSet& vtn_dependencies) {
+  assert(vtn->req() == n->req(), "scalars must have same number of reqs");
+  for (uint j = 0; j < n->req(); j++) {
+    Node* def = n->in(j);
+    if (def == nullptr) { continue; }
+    set_req_with_scalar(n, vtn, vtn_dependencies, j);
+  }
+}
+
+void SuperWordVTransformBuilder::set_all_req_with_vectors(const Node_List* pack, VTransformNode* vtn, VectorSet& vtn_dependencies) {
+  Node* p0 = pack->at(0);
+  assert(vtn->req() <= p0->req(), "must have at at most as many reqs");
+  // Vectors have no ctrl, so ignore it.
+  for (uint j = 1; j < vtn->req(); j++) {
+    Node* def = p0->in(j);
+    if (def == nullptr) { continue; }
+    set_req_with_vector(pack, vtn, vtn_dependencies, j);
+  }
+}
+
+void SuperWordVTransformBuilder::add_dependencies_of_node_to_vtn(Node*n, VTransformNode* vtn, VectorSet& vtn_dependencies) {
+  for (VLoopDependencyGraph::PredsIterator preds(_vloop_analyzer.dependency_graph(), n); !preds.done(); preds.next()) {
```

Comment:
Note: this `VLoopDependencyGraph` traversal used to be in `PacksetGraph::build`.